File: python/tests/test_amr_positivity_floor.py
#!/usr/bin/env python3
"""AmrSystem positivity_floor (Zhang-Shu, ADC-259): the public floor is now wired on the AMR transport.

Before ADC-259 the AMR facade REJECTED spatial.positivity_floor > 0 ("not supported on the AMR path").
It is now threaded facade -> AmrSystem::add_block -> advance_amr -> compute_face_fluxes (Density-role
face states) plus a Density clamp on the coarse-fine fine ghost means. This pure-Python test (no .so
compilation, always run) checks, on a native ISOTHERMAL block -- the ticket's target regime, p = cs2 rho,
where flooring the density also floors the pressure so face-density positivity is enough to keep the run
finite (compressible Euler can still drive the pressure negative independently, cf.
tests/test_positivity_floor.cpp; that is out of scope of the floor):

  (1) FACADE: positivity_floor > 0 no longer raises on the AMR path and the floored run stays FINITE on
      the 1e6-contrast oscillating top-hat advected at u=1 (where weno5 reconstructs a negative face
      density). The load-bearing claim (an unfloored run diverging on the same spike) is covered on the
      System path by tests/test_positivity_floor.cpp and on a refined AMR C/F interface by (3); it is not
      asserted here since it depended on the pre-ADC-324 never-tagged seed of set_refinement(1e30) (now a
      mono-level hierarchy), leaving a coarse-only grid on which the demo is not robust.
  (2) NO-DEFAULT-CHANGE: on a smooth positive drifting state, positivity_floor > 0 is BIT-IDENTICAL to
      positivity_floor = 0 (the floor never bites a face above it).
  (3) REGRID + MASS: with the floor active and a real refined hierarchy (the C/F interface straddling
      the 1e6 jump), the coarse mass is conserved across regrids and the state stays finite.
  (4) MULTI-BLOCK: positivity_floor rides the second facade path (build_multi -> dispatch_amr_block ->
      build_amr_block, AmrRuntime engine) too -- two floored blocks build and run finite.

The guarantee is face / C/F-ghost-mean Density positivity only (order-1 fallback), NOT updated-mean
positivity: the coarse MEAN density can still dip negative under aggressive CFL (parity with System).
The compiled .so AMR path REJECTS positivity_floor > 0 (flat ABI, no floor slot) -- covered by the
facade guards (python/adc/__init__.py, python/amr_system.cpp) and the C++ no-Density reject in
tests/test_amr_positivity_floor.cpp; not re-tested here to keep this test compiler-free.
"""
import sys

import numpy as np

import adc

CS2 = 0.25       # isothermal sound speed^2 (p = cs2 rho): flooring rho floors the pressure
n = 48
dt = 0.0008
fails = 0


def chk(cond, label):
    global fails
    print(f"  [{'OK ' if cond else 'XX '}] {label}")
    if not cond:
        fails += 1


def iso_spec():
    """Native isothermal block (3 var, Density role at component 0). No compiler required."""
    return adc.Model(state=adc.FluidState("isothermal", cs2=CS2), transport=adc.IsothermalFlux(),
                     source=adc.NoSource(), elliptic=adc.BackgroundDensity(alpha=0.0, n0=0.0))


def spike_state():
    """1e6-contrast top-hat in x (rho = 1 in the central band, 1e-6 outside) with a non-monotone
    oscillating spike at x ~ 3/4, advected at u = 1. rho[j, i] = density at cell (i, j); the band
    varies with i (x). set_conservative_state expects shape (ncomp, n, n)."""
    rho = np.full((n, n), 1e-6)
    rho[:, n // 3:2 * n // 3] = 1.0
    ks = 3 * n // 4
    rho[:, ks] = 0.8
    rho[:, ks + 1] = 0.5
    rho[:, ks + 2] = 1e-6
    rho[:, ks + 3] = 1.0
    rho[:, ks + 4] = 1e-6
    return np.stack([rho, rho * 1.0, 0.0 * rho])  # (3, n, n): u = 1 advects the contact


def smooth_state():
    """Smooth strictly-positive Gaussian bubble drifting at u = 0.3 (no face undershoots the floor)."""
    xs = (np.arange(n) + 0.5) / n
    X, Y = np.meshgrid(xs, xs, indexing="xy")
    rho = 1.0 + 0.4 * np.exp(-((X - 0.5) ** 2 + (Y - 0.5) ** 2) / 0.02)  # (n, n)
    return np.stack([rho, 0.3 * rho, 0.0 * rho])  # (3, n, n)


def build(pf, regrid_every=0, refine=1e30):
    s = adc.AmrSystem(n=n, L=1.0, periodic=True, regrid_every=regrid_every)
    s.set_refinement(refine)
    s.add_block("gas", iso_spec(),
                spatial=adc.Spatial(limiter="weno5", flux="rusanov", positivity_floor=pf),
                time=adc.Explicit())
    return s


def step_n(s, nsteps):
    for _ in range(nsteps):
        s.step(dt)


# --- (1) FACADE accepts the floor + the floored run stays finite on the 1e6-contrast spike ----------
# ADC-324: set_refinement(1e30) is now a MONO-LEVEL hierarchy (no seed fine patch). The historical
# "unfloored run blows up" assertion relied on the never-tagged 1e30 seed and is not robust on the
# resulting coarse-only grid (neither floored nor unfloored diverges there); the load-bearing property
# is covered on System by tests/test_positivity_floor.cpp and on a refined AMR C/F interface by (3).
# Here we keep the AMR-facade contract: floor>0 is accepted (previously raised at add_block) and the
# floored run stays finite over the spike.
print("== (1) positivity_floor>0 wired on AMR + floored run finite on the contrast-1e6 spike ==")
s_on = build(1e-8)                                  # previously raised ValueError at add_block
s_on.set_conservative_state("gas", spike_state())
step_n(s_on, 38)
d_on = np.asarray(s_on.density("gas"))
chk(np.all(np.isfinite(d_on)), "positivity_floor>0 accepted on AMR + finite over 38 steps")


# --- (2) NO-DEFAULT-CHANGE: smooth positive drift -> floor ON == floor OFF (bit-identical) ----------
print("== (2) no-default-change: floor>0 == floor=0 on smooth positive data ==")
sa = build(0.0)
sa.set_conservative_state("gas", smooth_state())
step_n(sa, 12)
sb = build(1e-8)
sb.set_conservative_state("gas", smooth_state())
step_n(sb, 12)
da, db = np.asarray(sa.density("gas")), np.asarray(sb.density("gas"))
chk(float(np.abs(da - db).max()) == 0.0, "floor>0 bit-identical to floor=0 on smooth state (dmax==0)")
chk(np.all(np.isfinite(db)), "floored smooth run finite")


# --- (3) REGRID + MASS: real refined hierarchy (C/F interface), floor active, mass conserved --------
print("== (3) floored AMR across a regrid: mass conserved, finite, fine patch created ==")
# refine threshold 0.5 sits between the 1e-6 background and the rho=1 band -> the band refines and the
# C/F interface straddles the 1e6 jump (the highest-risk reconstruction site).
s_rg = build(1e-8, regrid_every=4, refine=0.5)
s_rg.set_conservative_state("gas", spike_state())
m0 = s_rg.mass("gas")
step_n(s_rg, 16)
m1 = s_rg.mass("gas")
d_rg = np.asarray(s_rg.density("gas"))
chk(abs(m1 - m0) < 1e-9, "coarse mass conserved across regrid with floor (dm=%.2e)" % abs(m1 - m0))
chk(np.all(np.isfinite(d_rg)), "state finite across regrid with floor")
chk(s_rg.n_patches() >= 1, "fine patch created (C/F interface exercised)")


# --- (4) MULTI-BLOCK: positivity_floor threads dispatch_amr_block (AmrRuntime engine) ---------------
# Two blocks switch the single-block AmrCouplerMP for the multi-block AmrRuntime engine: the floor rides
# a DIFFERENT facade path (build_multi -> dispatch_amr_block -> build_amr_block) than (1)-(3).
# set_conservative_state is single-block only, so seed via set_density (u=0): the floor stays inactive
# but the path is accepted and runs finite -- a smoke test of the multi-block threading + arity.
print("== (4) multi-block: positivity_floor accepted + threaded through dispatch_amr_block ==")
band = np.full((n, n), 1e-6)
band[:, n // 3:2 * n // 3] = 1.0  # contrast-1e6 band (Density role, component 0)
sm = adc.AmrSystem(n=n, L=1.0, periodic=True)
sm.set_refinement(1e30)
sm.add_block("a", iso_spec(),
             spatial=adc.Spatial(limiter="weno5", flux="rusanov", positivity_floor=1e-8))
sm.add_block("b", iso_spec(),
             spatial=adc.Spatial(limiter="weno5", flux="rusanov", positivity_floor=1e-8))
sm.set_density("a", band.ravel().copy())
sm.set_density("b", band.ravel().copy())
step_n(sm, 5)
chk(np.all(np.isfinite(np.asarray(sm.density("a")))), "multi-block floor: block a finite")
chk(np.all(np.isfinite(np.asarray(sm.density("b")))), "multi-block floor: block b finite")


if fails:
    print(f"FAIL test_amr_positivity_floor : {fails} failure(s)")
    sys.exit(1)
print("OK test_amr_positivity_floor")

File: tests/test_amr_positivity_floor.cpp
// Zhang-Shu POSITIVITY FLOOR on the AMR transport path (ADC-259). The single-block uniform System
// floor (reconstruct_pp / zhang_shu_scale, tested in test_positivity_floor.cpp) is here exercised on
// the AMR engine: the floor is threaded facade -> advance_amr -> subcycle_level_mp ->
// compute_face_fluxes, AND a Density-role clamp is applied to the COARSE-FINE fine ghost means
// (fill_cf_ghost_cell, the refined-patch interface that the original Hoffart failure exercised).
//
// Four checks (serial; the np=1/2/4 MPI replay is the Python test, the device path is argued by
// construction -- same ADC_HD kernel, the clamp is host-side and outside every conservation sum):
//  (1) C/F GHOST CLAMP (the highest-risk site): the LIVE multi-box ghost fill mf_fill_fine_ghosts_mb
//      reads a coarse parent whose density dips BELOW the floor in one cell; unclamped the fine ghosts
//      reading it go sub-floor, with the floor every C/F fine-ghost density is >= floor while the
//      momenta stay interpolated (the clamp touches the Density role only).
//  (2) NO-DEFAULT-CHANGE: advance_amr on a SMOOTH positive Euler state is BIT-IDENTICAL with the floor
//      ON vs OFF (the floor never bites a face above it), and stays finite + mass-conserving.
//  (3) EFFECT: on the 1e6-contrast oscillatory top-hat (the profile where weno5 reconstructs a
//      negative face density) advance_amr WITH the floor stays finite and conserves mass, and its
//      trajectory DIFFERS from the unfloored run (the floor was active).
//  (4) REJECT: a model without a Density role (AdvectionDiffusion) + positivity_floor > 0 ->
//      runtime_error mentioning positivity_floor (positivity_comp resolved on the AMR path).
#include <adc/validation/physics/advection_diffusion.hpp>
#include <adc/physics/euler.hpp>

#include <adc/numerics/numerical_flux.hpp>      // RusanovFlux
#include <adc/numerics/reconstruction.hpp>      // Weno5, Minmod
#include <adc/numerics/spatial_operator.hpp>    // reconstruct_pp, positivity_comp
#include <adc/numerics/time/amr_reflux_mf.hpp>  // advance_amr, AmrLevelMP, mf_fill_fine_ghosts_mb

#include <adc/mesh/box2d.hpp>
#include <adc/mesh/box_array.hpp>
#include <adc/mesh/distribution_mapping.hpp>
#include <adc/mesh/for_each.hpp>
#include <adc/mesh/geometry.hpp>
#include <adc/mesh/multifab.hpp>
#include <adc/mesh/physical_bc.hpp>
#include <adc/parallel/comm.hpp>

#include <cmath>
#include <cstdio>
#include <stdexcept>
#include <string>

using namespace adc;

// Euler WITHOUT source (parity with test_positivity_floor.cpp): forwards flux / wave speed and the
// Density-role introspection (conservative_vars, role Density at component 0). advance_amr only needs
// flux / max_wave_speed / source / conservative_vars on the model.
struct EulerNoSrc {
  using State = Euler::State;
  using Aux = Euler::Aux;
  static constexpr int n_vars = Euler::n_vars;
  Euler e{};
  Real gamma = Real(1.4);
  ADC_HD State flux(const State& u, const Aux& a, int dir) const { return e.flux(u, a, dir); }
  ADC_HD Real max_wave_speed(const State& u, const Aux& a, int dir) const {
    return e.max_wave_speed(u, a, dir);
  }
  ADC_HD State source(const State&, const Aux&) const { return State{}; }
  static VariableSet conservative_vars() { return Euler::conservative_vars(); }
  static VariableSet primitive_vars() { return Euler::primitive_vars(); }
};

static constexpr double kRhoMin = 1e-6, kRhoMax = 1.0;  // contrast 1e6 (Hoffart top-hat)

// Smooth, strictly positive Euler bubble (no face ever undershoots the floor): used for (2).
static void init_smooth(MultiFab& U, const Geometry& geom, const EulerNoSrc& m) {
  for (int li = 0; li < U.local_size(); ++li) {
    Array4 a = U.fab(li).array();
    for_each_cell(U.box(li), [a, geom, m](int i, int j) {
      const Real x = geom.x_cell(i) - Real(0.5), y = geom.y_cell(j) - Real(0.5);
      const Real rho = Real(1.0) + Real(0.4) * std::exp(-(x * x + y * y) / Real(0.02));
      a(i, j, 0) = rho;
      a(i, j, 1) = Real(0.2) * rho;
      a(i, j, 2) = Real(-0.1) * rho;
      const Real ke = Real(0.5) * (a(i, j, 1) * a(i, j, 1) + a(i, j, 2) * a(i, j, 2)) / rho;
      a(i, j, 3) = Real(1.0) / (m.gamma - Real(1)) + ke;  // p = 1, plus kinetic energy
    });
  }
}

// 1e6-contrast top-hat in x with a non-monotone oscillating spike at x ~ 3/4 (the deterministic
// weno5 undershoot of test_positivity_floor.cpp): the EVOLVED contact reconstructs a negative face
// density. p uniform and tiny (quasi-cold, c ~ O(1) on the 1e-6 background). Used for (3).
static void init_spike(MultiFab& U, const Box2D& dom, const EulerNoSrc& m) {
  const int ilo = dom.hi[0] / 3, ihi = 2 * dom.hi[0] / 3;
  const int ks = 3 * dom.hi[0] / 4;
  const Real p0 = Real(1e-6);
  for (int li = 0; li < U.local_size(); ++li) {
    Array4 a = U.fab(li).array();
    for_each_cell(U.box(li), [a, ilo, ihi, ks, m, p0](int i, int j) {
      Real rho = (i >= ilo && i <= ihi) ? Real(kRhoMax) : Real(kRhoMin);
      if (i == ks) rho = Real(0.8);
      else if (i == ks + 1) rho = Real(0.5);
      else if (i == ks + 2) rho = Real(kRhoMin);
      else if (i == ks + 3) rho = Real(1.0);
      else if (i == ks + 4) rho = Real(kRhoMin);
      a(i, j, 0) = rho;
      a(i, j, 1) = rho * Real(1.0);  // u = 1: advected contact
      a(i, j, 2) = Real(0);
      a(i, j, 3) = p0 / (m.gamma - Real(1)) + Real(0.5) * rho;
    });
  }
}

int main(int argc, char** argv) {
  comm_init(&argc, &argv);
  const int me = my_rank();
  long fails = 0;
  auto chk = [&](bool c, const char* w) {
    if (!c) { if (me == 0) std::printf("FAIL %s\n", w); ++fails; }
  };

  std::printf("=== POSITIVITY FLOOR on AMR (positivity_floor, ADC-259) ===\n");

  const EulerNoSrc model{};
  const Real floor = Real(1e-6);
  const int n = 48;
  const Box2D dom = Box2D::from_extents(n, n);
  const Geometry geom{dom, 0.0, 1.0, 0.0, 1.0};
  const BoxArray ba(std::vector<Box2D>{dom});
  const DistributionMapping dm(1, n_ranks());
  const BCRec bc;  // periodic

  // ------------------------------------------------------------------------------------------------
  // (1) C/F GHOST CLAMP on the LIVE multi-box ghost fill (mf_fill_fine_ghosts_mb): a coarse parent
  //     with one sub-floor density cell -> the fine ghosts reading it dip below the floor unclamped,
  //     and are brought to >= floor (Density role only, momenta interpolated) with the floor.
  // ------------------------------------------------------------------------------------------------
  {
    const int nc = 8;
    const Box2D cdom = Box2D::from_extents(nc, nc);
    const BoxArray cba(std::vector<Box2D>{cdom});
    const DistributionMapping cdm(1, n_ranks());
    MultiFab Pc(cba, cdm, EulerNoSrc::n_vars, 0);
    Pc.set_val(0.0);
    if (Pc.local_size() > 0) {  // serial: coarse box is local; under np>1 the test is skipped clean
      sync_host();
      Array4 c = Pc.fab(0).array();
      for_each_cell(cdom, [c](int i, int j) {
        c(i, j, 0) = Real(1.0); c(i, j, 1) = Real(0.3); c(i, j, 2) = Real(-0.2); c(i, j, 3) = Real(2.0);
      });
      c(1, 3, 0) = Real(1e-10);  // sub-floor density (coarse means are NOT floored: reachable)
      c(1, 3, 1) = Real(0.5);    // distinctive momentum: the clamp must leave it untouched
    }
    // Fine patch [4,4]-[11,11] (covers coarse [2,2]-[5,5]); ng=2. The left ghosts at fine i in {2,3}
    // coarsen to ci=1; fine (3,6)/(3,7)/(2,6)/(2,7) coarsen to coarse (1,3) -> read the sub-floor cell.
    const Box2D fb{{4, 4}, {11, 11}};
    const BoxArray fba(std::vector<Box2D>{fb});
    const DistributionMapping fdm(1, n_ranks());
    MultiFab Uf(fba, fdm, EulerNoSrc::n_vars, 2);

    auto sub_floor_ghosts = [&](MultiFab& U) {
      int cnt = 0;
      if (U.local_size() > 0) {
        sync_host();
        const ConstArray4 f = U.fab(0).const_array();
        const Box2D v = U.box(0), g = U.fab(0).grown_box();
        for (int j = g.lo[1]; j <= g.hi[1]; ++j)
          for (int i = g.lo[0]; i <= g.hi[0]; ++i)
            if (!v.contains(i, j) && f(i, j, 0) < floor) ++cnt;
      }
      return cnt;
    };

    // unclamped (pos_floor = 0): the sub-floor coarse cell propagates to its fine ghosts.
    Uf.set_val(1.0);
    mf_fill_fine_ghosts_mb(Uf, Pc, Pc, Real(0.5), /*replicated_parent=*/true, Real(0), 0);
    const int n_raw = sub_floor_ghosts(Uf);

    // clamped (pos_floor = floor): every C/F fine-ghost density >= floor; momentum unchanged.
    Uf.set_val(1.0);
    mf_fill_fine_ghosts_mb(Uf, Pc, Pc, Real(0.5), /*replicated_parent=*/true, floor, 0);
    const int n_pp = sub_floor_ghosts(Uf);
    Real ghost_rho = Real(1), ghost_mom = Real(0);
    if (Uf.local_size() > 0) {
      sync_host();
      const ConstArray4 f = Uf.fab(0).const_array();
      ghost_rho = f(3, 6, 0);  // ghost reading coarse (1,3)
      ghost_mom = f(3, 6, 1);
    }
    if (me == 0)
      std::printf("(1) C/F ghost clamp : sub-floor ghosts NU = %d, PP = %d ; ghost(3,6) rho = %.3e "
                  "mom = %.3f\n", n_raw, n_pp, static_cast<double>(ghost_rho),
                  static_cast<double>(ghost_mom));
    if (n_ranks() == 1) {
      chk(n_raw > 0, "1_unclamped_ghost_subfloor");       // the problem exists at the C/F interface
      chk(n_pp == 0, "1_clamped_ghost_floor");            // the clamp fixes every C/F ghost
      chk(ghost_rho >= floor, "1_clamped_ghost_density_floored");
      chk(std::abs(ghost_mom - Real(0.5)) < Real(1e-12), "1_clamped_ghost_momentum_interpolated");
    }
  }

  // one explicit weno5 / rusanov AMR step on a single coarse level (leaf), in place.
  auto amr_step = [&](MultiFab& U, const MultiFab& aux, Real dt, Real pos_floor) {
    std::vector<AmrLevelMP> levels;
    levels.push_back(AmrLevelMP{U, &aux, geom.dx(), geom.dy()});
    advance_amr<Weno5, RusanovFlux>(model, levels, dom, dt, Periodicity{true, true},
                                    /*coarse_replicated=*/true, /*recon_prim=*/false, /*imex=*/false,
                                    NewtonOptions{}, AmrTimeMethod::kEuler, pos_floor);
    U = std::move(levels[0].U);
  };
  auto coarse_mass = [&](const MultiFab& U) {
    sync_host();
    double mass = 0;
    for (int li = 0; li < U.local_size(); ++li) {
      const ConstArray4 a = U.fab(li).const_array();
      const Box2D v = U.box(li);
      for (int j = v.lo[1]; j <= v.hi[1]; ++j)
        for (int i = v.lo[0]; i <= v.hi[0]; ++i) mass += a(i, j, 0);
    }
    return all_reduce_sum(mass);
  };
  auto finite_min_rho = [&](const MultiFab& U, bool& finite) {
    sync_host();
    Real mn = Real(1e30);
    finite = true;
    for (int li = 0; li < U.local_size(); ++li) {
      const ConstArray4 a = U.fab(li).const_array();
      const Box2D v = U.box(li);
      for (int j = v.lo[1]; j <= v.hi[1]; ++j)
        for (int i = v.lo[0]; i <= v.hi[0]; ++i) {
          mn = std::min(mn, a(i, j, 0));
          finite = finite && std::isfinite(static_cast<double>(a(i, j, 0)));
        }
    }
    return mn;
  };
  MultiFab aux(ba, dm, 3, 1);
  aux.set_val(0.0);

  // ------------------------------------------------------------------------------------------------
  // (2) NO-DEFAULT-CHANGE: smooth positive state -> advance_amr floor ON == floor OFF (bit-identical).
  // ------------------------------------------------------------------------------------------------
  {
    const int ng = Weno5::n_ghost;  // weno5 stencil radius (3)
    MultiFab U0(ba, dm, EulerNoSrc::n_vars, ng);
    init_smooth(U0, geom, model);
    const Real dt = Real(0.1) * geom.dx();
    MultiFab Uoff = U0, Uon = U0;
    amr_step(Uoff, aux, dt, Real(0));      // floor inactive
    amr_step(Uon, aux, dt, floor);         // floor active but never bites a face above it
    sync_host();
    double dmax = 0;
    for (int li = 0; li < Uoff.local_size(); ++li) {
      const ConstArray4 a = Uoff.fab(li).const_array(), b = Uon.fab(li).const_array();
      const Box2D v = Uoff.box(li);
      for (int j = v.lo[1]; j <= v.hi[1]; ++j)
        for (int i = v.lo[0]; i <= v.hi[0]; ++i)
          for (int cc = 0; cc < EulerNoSrc::n_vars; ++cc)
            dmax = std::max(dmax, std::abs(static_cast<double>(a(i, j, cc)) - b(i, j, cc)));
    }
    dmax = all_reduce_max(dmax);
    bool finite = true;
    finite_min_rho(Uon, finite);
    const double m0 = coarse_mass(U0), m1 = coarse_mass(Uon);
    if (me == 0)
      std::printf("(2) no-default-change : max|U(floor) - U(no floor)| = %.3e (smooth state)\n", dmax);
    chk(dmax == 0.0, "2_smooth_floor_bit_identical");
    chk(finite, "2_floor_finite");
    chk(std::abs(m1 - m0) < 1e-10, "2_mass_conserved");
  }

  // ------------------------------------------------------------------------------------------------
  // (3) EFFECT on the 1e6-contrast spike: advance_amr WITH the floor stays finite + conserves mass,
  //     and its trajectory differs from the unfloored run (the floor was active on the spike).
  // ------------------------------------------------------------------------------------------------
  {
    const int ng = Weno5::n_ghost;  // weno5 stencil radius (3)
    MultiFab U0(ba, dm, EulerNoSrc::n_vars, ng);
    init_spike(U0, dom, model);
    const Real dt = Real(0.05) * geom.dx() / Real(2.5);  // prudent CFL (u=1, c~1.2 on the background)
    const double m0 = coarse_mass(U0);
    MultiFab Uon = U0, Uoff = U0;
    amr_step(Uon, aux, dt, floor);
    amr_step(Uoff, aux, dt, Real(0));
    bool finite_on = true;
    finite_min_rho(Uon, finite_on);
    const double m1 = coarse_mass(Uon);
    // "active": the floored and unfloored trajectories differ on a valid cell (NaN counts as differ).
    sync_host();
    bool differs = false;
    for (int li = 0; li < Uon.local_size(); ++li) {
      const ConstArray4 a = Uon.fab(li).const_array(), b = Uoff.fab(li).const_array();
      const Box2D v = Uon.box(li);
      for (int j = v.lo[1]; j <= v.hi[1]; ++j)
        for (int i = v.lo[0]; i <= v.hi[0]; ++i)
          for (int cc = 0; cc < EulerNoSrc::n_vars; ++cc) {
            const double da = a(i, j, cc), db = b(i, j, cc);
            if (!(da == db)) differs = true;  // NaN != NaN -> differ
          }
    }
    differs = all_reduce_max(differs ? 1.0 : 0.0) > 0.0;
    if (me == 0)
      std::printf("(3) spike + floor : finite = %d, mass drift = %.3e, trajectory differs = %d\n",
                  finite_on ? 1 : 0, m1 - m0, differs ? 1 : 0);
    chk(finite_on, "3_floored_spike_finite");
    chk(std::abs(m1 - m0) < 1e-10, "3_floored_spike_mass_conserved");
    chk(differs, "3_floor_active_on_spike");
  }

  // ------------------------------------------------------------------------------------------------
  // (4) REJECT: a model without a Density role + positivity_floor > 0 -> runtime_error on the AMR path
  //     (positivity_comp resolved at the head of subcycle_level_mp), message mentions positivity_floor.
  // ------------------------------------------------------------------------------------------------
  {
    const adc::validation::AdvectionDiffusion scal{1.0, 0.0, 0.0};
    MultiFab Us(ba, dm, 1, 2), auxs(ba, dm, 3, 1);
    Us.set_val(1.0);
    auxs.set_val(0.0);
    bool threw = false;
    std::string msg;
    try {
      std::vector<AmrLevelMP> levels;
      levels.push_back(AmrLevelMP{Us, &auxs, geom.dx(), geom.dy()});
      advance_amr<Minmod, RusanovFlux>(scal, levels, dom, Real(1e-4), Periodicity{true, true},
                                       /*coarse_replicated=*/true, /*recon_prim=*/false,
                                       /*imex=*/false, NewtonOptions{}, AmrTimeMethod::kEuler,
                                       /*pos_floor=*/floor);
    } catch (const std::runtime_error& e) {
      threw = true;
      msg = e.what();
    }
    if (me == 0)
      std::printf("(4) model without Density : %s (%s)\n", threw ? "REJECTED" : "ACCEPTED (!)",
                  msg.c_str());
    chk(threw, "4_reject_model_without_density");
    chk(msg.find("positivity_floor") != std::string::npos, "4_message_mentions_positivity_floor");
  }

  fails = static_cast<long>(all_reduce_max(static_cast<double>(fails)));
  if (me == 0 && fails == 0) std::printf("OK test_amr_positivity_floor\n");
  comm_finalize();
  return fails == 0 ? 0 : 1;
}
